### ISSUE TITLE
GPy (and possibly others) imported before being installed

### ISSUE THREAD

**Issue by sigvaldm** [mathLab/EZyRB#124]
setup.py imports ezyrb, which imports GPy (and others). Although setup.py lists GPy as a requirement, it may not be installed at the time setup.py is run, which leads to the following error upon `pip install .`

```
    ERROR: Command errored out with exit status 1:
     command: /usr/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/home/sigvald/Codes/EZyRB/setup.py'"'"'; __file__='"'"'/home/sigvald/Codes/EZyRB/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-6bkra3yh
         cwd: /home/sigvald/Codes/EZyRB/
    Complete output (9 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/home/sigvald/Codes/EZyRB/setup.py", line 4, in <module>
        import ezyrb
      File "/home/sigvald/Codes/EZyRB/ezyrb/__init__.py", line 23, in <module>
        from .gpr import GPR
      File "/home/sigvald/Codes/EZyRB/ezyrb/gpr.py", line 4, in <module>
        import GPy
    ModuleNotFoundError: No module named 'GPy'
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
```

Some relevant reading: https://packaging.python.org/guides/single-sourcing-package-version/

**Comment by ndem0** [comment]
Dear @sigvaldm, thanks for the feedback. I do not completely get your error: have you tried to install *EZyRB* by using `pip` or through the `setup.py` file?

**Comment by sigvaldm** [comment]
I git cloned the library to a local folder, and ran `pip install .` inside that folder (mind the `.` specifying the local directory). `pip` uses `setup.py`. Does this answer your question?

**Comment by sigvaldm** [comment]
Just to verify once more, I tried also on a fresh Conda environment (full report below). This time I get the same with numpy. Numpy is imported by setup.py before it is installed. This does not happen if you do `pip install ezyrb`, installing EZyRB from PyPI. I believe this is because the version on PyPI, EZyRB 1.0, do not import ezyrb in setup.py. If you deployed v1.1 to PyPI I believed you would get the same error there. You don't get it in Travis CI because there you install the dependencies manually, instead of letting pip do the work for you. Makes sense?

```
[sigvald@sim EZyRB]$ conda create -n ezyrb python=3
Collecting package metadata (current_repodata.json): done
Solving environment: done


==> WARNING: A newer version of conda exists. <==
  current version: 4.7.5
  latest version: 4.9.2

Please update conda by running

    $ conda update -n base -c defaults conda



## Package Plan ##

  environment location: /home/sigvald/.conda/envs/ezyrb

  added / updated specs:
    - python=3


The following NEW packages will be INSTALLED:

  _libgcc_mutex      pkgs/main/linux-64::_libgcc_mutex-0.1-main
  ca-certificates    pkgs/main/linux-64::ca-certificates-2020.12.8-h06a4308_0
  certifi            pkgs/main/linux-64::certifi-2020.12.5-py39h06a4308_0
  ld_impl_linux-64   pkgs/main/linux-64::ld_impl_linux-64-2.33.1-h53a641e_7
  libedit            pkgs/main/linux-64::libedit-3.1.20191231-h14c3975_1
  libffi             pkgs/main/linux-64::libffi-3.3-he6710b0_2
  libgcc-ng          pkgs/main/linux-64::libgcc-ng-9.1.0-hdf63c60_0
  libstdcxx-ng       pkgs/main/linux-64::libstdcxx-ng-9.1.0-hdf63c60_0
  ncurses            pkgs/main/linux-64::ncurses-6.2-he6710b0_1
  openssl            pkgs/main/linux-64::openssl-1.1.1i-h27cfd23_0
  pip                pkgs/main/linux-64::pip-20.3.1-py39h06a4308_0
  python             pkgs/main/linux-64::python-3.9.1-hdb3f193_2
  readline           pkgs/main/linux-64::readline-8.0-h7b6447c_0
  setuptools         pkgs/main/linux-64::setuptools-51.0.0-py39h06a4308_2
  sqlite             pkgs/main/linux-64::sqlite-3.33.0-h62c20be_0
  tk                 pkgs/main/linux-64::tk-8.6.10-hbc83047_0
  tzdata             pkgs/main/noarch::tzdata-2020d-h14c3975_0
  wheel              pkgs/main/noarch::wheel-0.36.2-pyhd3eb1b0_0
  xz                 pkgs/main/linux-64::xz-5.2.5-h7b6447c_0
  zlib               pkgs/main/linux-64::zlib-1.2.11-h7b6447c_3


Proceed ([y]/n)? 

Preparing transaction: done
Verifying transaction: done
Executing transaction: done
#
# To activate this environment, use
#
#     $ conda activate ezyrb
#
# To deactivate an active environment, use
#
#     $ conda deactivate

  7,37s user 1,02s system 79% cpu 10,498 total
[sigvald@sim EZyRB]$ conda activate ezyrb
(ezyrb) [sigvald@sim EZyRB]$ git checkout v1.1
HEAD is now at 1d7cf1b Action version
(ezyrb) [sigvald@sim EZyRB]$ pip install .
Processing /home/sigvald/Codes/EZyRB
    ERROR: Command errored out with exit status 1:
     command: /home/sigvald/.conda/envs/ezyrb/bin/python -c 'import sys, setuptools, tokenize; sys.argv[0] = '"'"'/tmp/pip-req-build-sk0r3jf9/setup.py'"'"'; __file__='"'"'/tmp/pip-req-build-sk0r3jf9/setup.py'"'"';f=getattr(tokenize, '"'"'open'"'"', open)(__file__);code=f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' egg_info --egg-base /tmp/pip-pip-egg-info-oy96x831
         cwd: /tmp/pip-req-build-sk0r3jf9/
    Complete output (9 lines):
    Traceback (most recent call last):
      File "<string>", line 1, in <module>
      File "/tmp/pip-req-build-sk0r3jf9/setup.py", line 4, in <module>
        import ezyrb
      File "/tmp/pip-req-build-sk0r3jf9/ezyrb/__init__.py", line 17, in <module>
        from .database import Database
      File "/tmp/pip-req-build-sk0r3jf9/ezyrb/database.py", line 4, in <module>
        import numpy as np
    ModuleNotFoundError: No module named 'numpy'
    ----------------------------------------
ERROR: Command errored out with exit status 1: python setup.py egg_info Check the logs for full command output.
(ezyrb) [sigvald@sim EZyRB]$ 
```

**Comment by ndem0** [comment]
Ok, now I've understood. Yes, you're right, it is a little bug. Thanks again for the effort (also for double-checking) and for the proposed solution: I'll check that the modification will not break the documentation deploy and I'll merge your PR.

**Comment by sigvaldm** [comment]
Great! Let me know if you'd like me to change something.

**Comment by ndem0** [comment]
Fixed in #125
